File: code/src/isalsr/core/native/src/string_to_dag.cpp
/// string_to_dag.cpp — NativeStringToDAG implementation.
///
/// Matches Python's StringToDAG (string_to_dag.py) token-for-token.
/// All Critical Invariants are documented at each decision point.

#include <isalsr/string_to_dag.hpp>

#include <cmath>
#include <cstdint>
#include <limits>
#include <stdexcept>
#include <string>
#include <vector>

namespace isalsr {

// ============================================================================
// Tokenizer
// ============================================================================

std::vector<std::string> tokenize(const std::string& input_string) {
    std::vector<std::string> tokens;
    const std::size_t n = input_string.size();
    std::size_t i = 0;

    while (i < n) {
        const char ch = input_string[i];

        if (ch == 'V' || ch == 'v') {
            // Compound token: consume the next character as a label.
            // A trailing V/v with no following character is invalid.
            if (i + 1 >= n) {
                throw std::invalid_argument(
                    std::string("'") + ch +
                    "' at position " + std::to_string(i) +
                    " requires a label character, but the string ends"
                );
            }
            const char label = input_string[i + 1];
            if (!is_valid_label_char(label)) {
                throw std::invalid_argument(
                    std::string("Invalid label character '") + label +
                    "' at position " + std::to_string(i + 1) +
                    " after '" + ch + "'"
                );
            }
            // Two-character token, e.g. "V+" or "vc".
            tokens.push_back(std::string{ch, label});
            i += 2;

        } else if (is_single_instruction(ch)) {
            // Single-character instruction: N P n p C c W.
            tokens.push_back(std::string{ch});
            i += 1;

        } else {
            throw std::invalid_argument(
                std::string("Invalid character '") + ch +
                "' at position " + std::to_string(i)
            );
        }
    }

    return tokens;
}

// ============================================================================
// NativeStringToDAG
// ============================================================================

NativeStringToDAG::NativeStringToDAG(
    const std::string& input_string,
    int32_t            num_variables
)
    : input_string_(input_string)
    , num_variables_(num_variables)
    , tokens_(tokenize(input_string))
{
    if (num_variables < 1) {
        throw std::invalid_argument(
            "num_variables must be >= 1, got " + std::to_string(num_variables)
        );
    }
}

NativeLabeledDAG NativeStringToDAG::run() const {
    // Pre-compute capacity: m variables + count of V/v two-char insertion tokens.
    // Invariant 7: variables are pre-inserted, not created by V/v.
    int32_t insertion_count = 0;
    for (const auto& t : tokens_) {
        if (t.size() == 2) ++insertion_count;
    }
    const int32_t max_nodes = num_variables_ + insertion_count;

    NativeLabeledDAG dag(max_nodes);
    NativeCDLL       cdll(max_nodes);

    // ---- Initial state: m VAR nodes, both pointers on x_1's CDLL slot ----
    // Invariant 7: Insert m VAR nodes before executing any instruction.
    int32_t prev_cdll_node  = -1;   // -1 → CDLL is empty; insert_after ignores it
    int32_t first_cdll_node = -1;

    static constexpr double kNaN = std::numeric_limits<double>::quiet_NaN();

    for (int32_t i = 0; i < num_variables_; ++i) {
        // Graph node IDs are assigned sequentially by add_node.
        const int32_t graph_node = dag.add_node(NodeType::VAR, i, kNaN);
        const int32_t cdll_node  = cdll.insert_after(prev_cdll_node, graph_node);
        if (i == 0) first_cdll_node = cdll_node;
        prev_cdll_node = cdll_node;
    }

    // Both pointers start on x_1's CDLL node (Invariant 7).
    int32_t primary_ptr   = first_cdll_node;
    int32_t secondary_ptr = first_cdll_node;

    // ---- Execute each token ----
    for (const auto& token : tokens_) {
        execute_token(token, dag, cdll, primary_ptr, secondary_ptr);
    }

    return dag;   // returned by value; caller owns an independent copy
}

// ============================================================================
// Private helpers
// ============================================================================

void NativeStringToDAG::execute_token(
    const std::string& token,
    NativeLabeledDAG&  dag,
    NativeCDLL&        cdll,
    int32_t&           primary_ptr,
    int32_t&           secondary_ptr
) {
    if (token == "N") {
        primary_ptr = cdll.next_node(primary_ptr);

    } else if (token == "P") {
        primary_ptr = cdll.prev_node(primary_ptr);

    } else if (token == "n") {
        secondary_ptr = cdll.next_node(secondary_ptr);

    } else if (token == "p") {
        secondary_ptr = cdll.prev_node(secondary_ptr);

    } else if (token == "C") {
        // Edge: primary → secondary.  DAG cycle check via add_edge; silent no-op on cycle.
        // Invariant 6: if add_edge would create a cycle, it returns false and we ignore it.
        const int32_t src = cdll.get_value(primary_ptr);
        const int32_t tgt = cdll.get_value(secondary_ptr);
        dag.add_edge(src, tgt);

    } else if (token == "c") {
        // Edge: secondary → primary.  Same cycle-check semantics.
        const int32_t src = cdll.get_value(secondary_ptr);
        const int32_t tgt = cdll.get_value(primary_ptr);
        dag.add_edge(src, tgt);

    } else if (token == "W") {
        // No-op.

    } else if (token.size() == 2 && (token[0] == 'V' || token[0] == 'v')) {
        execute_insertion(token, dag, cdll, primary_ptr, secondary_ptr);

    } else {
        // Should not reach here if the tokenizer is correct.
        throw std::invalid_argument(
            std::string("Unknown token: ") + token
        );
    }
}

void NativeStringToDAG::execute_insertion(
    const std::string& token,
    NativeLabeledDAG&  dag,
    NativeCDLL&        cdll,
    int32_t&           primary_ptr,
    int32_t&           secondary_ptr
) {
    const char pointer_char = token[0];  // 'V' or 'v'
    const char label_char   = token[1];
    const NodeType node_type = label_char_to_node_type(label_char);

    static constexpr double kNaN      = std::numeric_limits<double>::quiet_NaN();
    static constexpr double kConstOne = 1.0;  // matches Python: const_value=1.0

    // Create the new node.
    // CONST nodes receive initial const_value=1.0 (Python: StringToDAG._execute_insertion).
    // All other non-VAR nodes receive no metadata (-1, NaN).
    const double  cv       = (node_type == NodeType::CONST) ? kConstOne : kNaN;
    const int32_t new_node = dag.add_node(node_type, -1, cv);

    if (pointer_char == 'V') {
        // Edge: primary's graph node → new node.  Invariant 3: V/v sets first edge.
        const int32_t primary_graph = cdll.get_value(primary_ptr);
        dag.add_edge(primary_graph, new_node);
        // Insert new node into CDLL after primary.
        // Invariant 4: primary_ptr does NOT advance after insertion.
        cdll.insert_after(primary_ptr, new_node);
    } else {
        // pointer_char == 'v': same semantics using secondary pointer.
        const int32_t secondary_graph = cdll.get_value(secondary_ptr);
        dag.add_edge(secondary_graph, new_node);
        // Invariant 4: secondary_ptr does NOT advance after insertion.
        cdll.insert_after(secondary_ptr, new_node);
    }
}

} // namespace isalsr

File: code/src/isalsr/core/string_to_dag.py
"""StringToDAG (S2D) -- Execute an IsalSR instruction string to produce a LabeledDAG.

Tokenizes the input string into IsalSR tokens and executes them sequentially,
building a LabeledDAG. Handles the two-tier encoding: single-char tokens for
movement/edge/no-op, and two-char compound tokens for labeled node insertion.

Reference: IsalGraph's StringToGraph at
    /home/mpascual/research/code/IsalGraph/src/isalgraph/core/string_to_graph.py

Restriction: ZERO external dependencies. Only Python stdlib.
"""

from __future__ import annotations

import logging
from copy import deepcopy

from isalsr.core.cdll import CircularDoublyLinkedList
from isalsr.core.labeled_dag import LabeledDAG
from isalsr.core.node_types import LABEL_CHAR_MAP, NodeType, OperationSet
from isalsr.errors import InvalidTokenError
from isalsr.types import VALID_LABEL_CHARS, VALID_SINGLE_INSTRUCTIONS, InstructionToken

log = logging.getLogger(__name__)

# Trace entry: (dag_snapshot, cdll_snapshot, primary_ptr, secondary_ptr, tokens_so_far)
TraceEntry = tuple[LabeledDAG, CircularDoublyLinkedList, int, int, list[InstructionToken]]


class StringToDAG:
    """Convert an IsalSR instruction string into a ``LabeledDAG``.

    Args:
        input_string: The instruction string using the IsalSR two-tier encoding.
        num_variables: Number of input variables (m). Creates m VAR nodes as initial state.
        allowed_ops: If provided, restricts which operations V/v can create.
            Tokens for disallowed operations raise ``InvalidTokenError``.

    Raises:
        InvalidTokenError: If the string contains invalid tokens or disallowed operations.
        ValueError: If ``num_variables`` < 1.
    """

    __slots__ = (
        "_input_string",
        "_num_variables",
        "_allowed_ops",
        "_tokens",
        "_max_nodes",
        "_output_dag",
        "_cdll",
        "_primary_ptr",
        "_secondary_ptr",
        "_trace_log",
    )

    def __init__(
        self,
        input_string: str,
        num_variables: int,
        allowed_ops: OperationSet | None = None,
    ) -> None:
        if num_variables < 1:
            raise ValueError(f"num_variables must be >= 1, got {num_variables}")

        self._input_string: str = input_string
        self._num_variables: int = num_variables
        self._allowed_ops: OperationSet | None = allowed_ops

        # Tokenize and validate.
        self._tokens: list[InstructionToken] = _tokenize(input_string, allowed_ops)

        # Pre-compute capacity: m variables + count of V/v insertion tokens.
        insertion_count = sum(1 for t in self._tokens if len(t) == 2)
        self._max_nodes: int = num_variables + insertion_count

        # Pre-allocate data structures.
        self._output_dag: LabeledDAG = LabeledDAG(self._max_nodes)
        self._cdll: CircularDoublyLinkedList = CircularDoublyLinkedList(self._max_nodes)
        self._primary_ptr: int = -1
        self._secondary_ptr: int = -1
        self._trace_log: list[TraceEntry] = []

    # ------------------------------------------------------------------
    # Public accessors (useful for trace / debugging)
    # ------------------------------------------------------------------

    @property
    def tokens(self) -> list[InstructionToken]:
        """The tokenized instruction list."""
        return list(self._tokens)

    @property
    def cdll(self) -> CircularDoublyLinkedList:
        """The CDLL after (or during) conversion."""
        return self._cdll

    @property
    def primary_ptr(self) -> int:
        """Current primary pointer (CDLL node index)."""
        return self._primary_ptr

    @property
    def secondary_ptr(self) -> int:
        """Current secondary pointer (CDLL node index)."""
        return self._secondary_ptr

    # ------------------------------------------------------------------
    # Conversion
    # ------------------------------------------------------------------

    def run(self, *, trace: bool = False) -> LabeledDAG:
        """Execute the string-to-DAG conversion.

        Args:
            trace: If ``True``, collect deep-copied snapshots after each token.
                Snapshots are stored in ``self.trace_log``.

        Returns:
            The resulting LabeledDAG.
        """
        # ---- Initial state: m VAR nodes, both pointers on x_1 ----
        self._initialize_variables()

        trace_log: list[TraceEntry] = []
        if trace:
            trace_log.append(self._snapshot([]))

        # ---- Process each token ----
        for idx, token in enumerate(self._tokens):
            self._execute_token(token)
            if trace:
                trace_log.append(self._snapshot(self._tokens[: idx + 1]))

        # Store trace for external access.
        self._trace_log = trace_log
        return self._output_dag

    # ------------------------------------------------------------------
    # Initialization
    # ------------------------------------------------------------------

    def _initialize_variables(self) -> None:
        """Create m VAR nodes and insert them into the CDLL in order."""
        prev_cdll_node: int = -1
        first_cdll_node: int = -1

        for i in range(self._num_variables):
            graph_node = self._output_dag.add_node(NodeType.VAR, var_index=i)
            cdll_node = self._cdll.insert_after(prev_cdll_node, graph_node)
            if i == 0:
                first_cdll_node = cdll_node
            prev_cdll_node = cdll_node

        # Both pointers on the CDLL node for x_1 (first variable).
        self._primary_ptr = first_cdll_node
        self._secondary_ptr = first_cdll_node

    # ------------------------------------------------------------------
    # Instruction dispatch
    # ------------------------------------------------------------------

    def _execute_token(self, token: InstructionToken) -> None:
        """Execute a single IsalSR token, mutating internal state."""
        if token == "N":
            self._primary_ptr = self._cdll.next_node(self._primary_ptr)

        elif token == "P":
            self._primary_ptr = self._cdll.prev_node(self._primary_ptr)

        elif token == "n":
            self._secondary_ptr = self._cdll.next_node(self._secondary_ptr)

        elif token == "p":
            self._secondary_ptr = self._cdll.prev_node(self._secondary_ptr)

        elif token == "C":
            # Edge: primary -> secondary (with DAG cycle check).
            src = self._cdll.get_value(self._primary_ptr)
            tgt = self._cdll.get_value(self._secondary_ptr)
            # add_edge returns False if cycle/duplicate; silently skip.
            self._output_dag.add_edge(src, tgt)

        elif token == "c":
            # Edge: secondary -> primary (with DAG cycle check).
            src = self._cdll.get_value(self._secondary_ptr)
            tgt = self._cdll.get_value(self._primary_ptr)
            self._output_dag.add_edge(src, tgt)

        elif token == "W":
            pass  # No-op.

        elif len(token) == 2 and token[0] in "Vv":
            self._execute_insertion(token)

        else:
            # Should not reach here if tokenizer is correct, but be defensive.
            raise InvalidTokenError(f"Unknown token: {token!r}")

    def _execute_insertion(self, token: InstructionToken) -> None:
        """Execute a V[label] or v[label] insertion token."""
        pointer_char = token[0]
        label_char = token[1]
        node_type = LABEL_CHAR_MAP[label_char]

        # Create the new node.
        if node_type == NodeType.CONST:
            new_node = self._output_dag.add_node(node_type, const_value=1.0)
        else:
            new_node = self._output_dag.add_node(node_type)

        if pointer_char == "V":
            # Edge from primary's graph node to new node.
            primary_graph_node = self._cdll.get_value(self._primary_ptr)
            self._output_dag.add_edge(primary_graph_node, new_node)
            # Insert new node into CDLL after primary. Pointer does NOT move.
            self._cdll.insert_after(self._primary_ptr, new_node)
        else:
            # pointer_char == "v": same using secondary pointer.
            secondary_graph_node = self._cdll.get_value(self._secondary_ptr)
            self._output_dag.add_edge(secondary_graph_node, new_node)
            self._cdll.insert_after(self._secondary_ptr, new_node)

    # ------------------------------------------------------------------
    # Trace support
    # ------------------------------------------------------------------

    def _snapshot(self, tokens_so_far: list[InstructionToken]) -> TraceEntry:
        """Create a deep-copied snapshot of the current state."""
        return (
            deepcopy(self._output_dag),
            deepcopy(self._cdll),
            self._primary_ptr,
            self._secondary_ptr,
            list(tokens_so_far),
        )


# ======================================================================
# Module-level tokenizer
# ======================================================================


def _tokenize(
    input_string: str,
    allowed_ops: OperationSet | None = None,
) -> list[InstructionToken]:
    """Tokenize an IsalSR instruction string into a list of tokens.

    Tokenizer rules:
        - If current char is 'V' or 'v', consume the next char as the label.
          The compound token is the two chars together (e.g., 'V+', 'vs').
        - If 'V'/'v' is at end of string with no next char, raise InvalidTokenError.
        - Otherwise, the single char is a token (N, P, n, p, C, c, W).
        - Unknown characters raise InvalidTokenError.
        - If allowed_ops is set, validate label chars against allowed_ops.label_chars.

    Args:
        input_string: The raw IsalSR instruction string.
        allowed_ops: If provided, restricts which label chars are valid.

    Returns:
        List of InstructionToken (single-char or two-char strings).

    Raises:
        InvalidTokenError: On invalid characters, truncated V/v, or disallowed ops.
    """
    tokens: list[InstructionToken] = []
    i = 0
    n = len(input_string)

    while i < n:
        ch = input_string[i]

        if ch in "Vv":
            # Compound token: consume next char as label.
            if i + 1 >= n:
                raise InvalidTokenError(
                    f"'{ch}' at position {i} requires a label character, but the string ends"
                )
            label = input_string[i + 1]
            if label not in VALID_LABEL_CHARS:
                raise InvalidTokenError(
                    f"Invalid label character '{label}' at position {i + 1} "
                    f"after '{ch}'. Valid labels: {sorted(VALID_LABEL_CHARS)}"
                )
            if allowed_ops is not None and label not in allowed_ops.label_chars:
                raise InvalidTokenError(
                    f"Operation '{label}' (token '{ch}{label}') is not in the allowed operation set"
                )
            tokens.append(ch + label)
            i += 2

        elif ch in VALID_SINGLE_INSTRUCTIONS:
            tokens.append(ch)
            i += 1

        else:
            raise InvalidTokenError(
                f"Invalid character '{ch}' at position {i}. "
                f"Expected one of {sorted(VALID_SINGLE_INSTRUCTIONS)} or V/v"
            )

    return tokens
